gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -190,18 +191,160 @@ public boolean isJobLockEnabled() {
   /**
    * Get the job state of the most recent run of the job.
    */
-  private JobState getPreviousJobState(String jobName)
-      throws IOException {
-    if (this.jobStateStore.exists(jobName, "current" + AbstractJobLauncher.JOB_STATE_STORE_TABLE_SUFFIX)) {
-      // Read the job state of the most recent run of the job
-      List<JobState> previousJobStateList =
-          this.jobStateStore.getAll(jobName, "current" + AbstractJobLauncher.JOB_STATE_STORE_TABLE_SUFFIX);
-      if (!previousJobStateList.isEmpty()) {
-        // There should be a single job state on the list if the list is not empty
-        return previousJobStateList.get(0);
+  private Map<String, JobState> getPreviousDatasetStatesByUrns(String jobName) throws IOException {
+    Path stateStorePath = new Path(this.stateStoreRootDir, jobName);
+    if (!this.stateStoreFs.exists(stateStorePath)) {
+      return ImmutableMap.of();
+    }
+
+    FileStatus[] stateStoreFileStatuses = this.stateStoreFs.listStatus(stateStorePath, new PathFilter() {
```

Comment:
I agree on this. Actually I think the entire `StateStore` interface needs some major refactoring to make it more usable and less-prone to misuse. I personally think it's better to work on that in a separate PR because this one is already pretty big. What do you think?